### PR TITLE
[SPARK-12400][Shuffle] Avoid generating temp shuffle files for empty partitions

### PR DESCRIPTION
**Pull request by jerryshao**
This problem lies in `BypassMergeSortShuffleWriter`, empty partition will also generate a temp shuffle file with several bytes. So here change to only create file when partition is not empty.

This problem only lies in here, no such issue in `HashShuffleWriter`.

Please help to review, thanks a lot.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Hey, just curious: did this result in any perf. improvements? I've considered this change a couple of times but in my own benchmarking work it didn't seem to make a huge difference.

**Comment by rxin** [comment]
This reduces the number of files when using a very large number of reducers with little data. Good to do unless there are major risks.

**Comment by jerryshao** [comment]
Hi @JoshRosen , from performance point I don't think there's a big difference with this patch, since at most we will only open `200 * Cores` number of files simultaneously. But at least we could avoid generating file while the related partition is empty.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by jerryshao** [comment]
Jenkins, retest this please.

**Comment by jerryshao** [comment]
Jenkins, retest this please.

**Comment by jerryshao** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

File: core/src/test/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriterSuite.scala [hunk 1]
```diff
@@ -169,6 +170,41 @@ class BypassMergeSortShuffleWriterSuite extends SparkFunSuite with BeforeAndAfte
     assert(taskMetrics.memoryBytesSpilled === 0)
   }
 
+  test("only generate temp shuffle file for non-empty partition") {
+    // Using exception to test whether only non-empty partition creates temp shuffle file,
```
Comment (JoshRosen) [inline core/src/test/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriterSuite.scala]: `temporaryFilesCreated` is only cleared in `afterAll()`, so it should contain every file which was created even after those files are deleted from the disk. Thus, I think we can simplify this test a bit by not relying on the exception here: just check that `temporaryFilesCreated.size === 3` while `temporaryFilesCreated.count(_.exists()) === 0` after a successful stop.

File: core/src/test/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriterSuite.scala [hunk 2]
```diff
@@ -104,7 +104,7 @@ class BypassMergeSortShuffleWriterSuite extends SparkFunSuite with BeforeAndAfte
       new Answer[(TempShuffleBlockId, File)] {
         override def answer(invocation: InvocationOnMock): (TempShuffleBlockId, File) = {
           val blockId = new TempShuffleBlockId(UUID.randomUUID)
-          val file = File.createTempFile(blockId.toString, null, tempDir)
+          val file = new File(tempDir, blockId.name)
```
Comment (JoshRosen) [inline core/src/test/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriterSuite.scala]: Why is this change necessary?

**Comment by JoshRosen** [comment]
This looks good to me overall. I have a couple of suggestions for how we might simplify the test case; please take a look at my comments and let me know whether you agree with them. This is fine to merge in its current form (pending Jenkins) but I wouldn't mind simplifying the test if it's possible. Thanks!

File: core/src/test/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriterSuite.scala (hunk 2 above)
Comment (jerryshao) [inline core/src/test/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriterSuite.scala]: Because `File.createTempFile()` will create an empty file even if you don't use it, so this will break the current changes.

**Comment by jerryshao** [comment]
Thanks a lot @JoshRosen for your comments, I will try to simplify the test if possible.

File: core/src/test/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriterSuite.scala (hunk 2 above)
Comment (JoshRosen) [inline core/src/test/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriterSuite.scala]: Ah, gotcha. Makes sense.

File: core/src/test/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriterSuite.scala (hunk 1 above)
Comment (JoshRosen) [inline core/src/test/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriterSuite.scala]: Ah, wait: I see now that this won't work due to when `createTempShuffleBlock()` is called: it's invoked before the `open()`. Therefore the current code looks fine to me.

**Comment by JoshRosen** [comment]
Actually, upon closer inspection I think that my suggestion might not work because of the lifecycle of when these methods are called. Therefore I'm now inclined to merge this as-is; we can do a larger test cleanup later if we'd like to.

**Comment by JoshRosen** [comment]
Merging this into master (2.0.0). Thanks @jerryshao!
